File: patient-ios-app/MediHubPatient/Core/Models/AppointmentModels.swift
import Foundation

// MARK: - Appointment Models

struct AppointmentDTO: Codable, Identifiable, Hashable {
    let id: String?
    let patientId: String?
    let patientName: String?
    let patientEmail: String?
    let patientPhone: String?
    let staffId: String?
    /// The clinician's USER id — what /chat/send needs as a recipient.
    /// `staffId` is a Staff row id and is not interchangeable with it.
    /// Android has always decoded this; iOS did not, which is why the
    /// composer had no way to address anyone.
    let staffUserId: String?
    let staffName: String?
    let staffEmail: String?
    let hospitalId: String?
    let hospitalName: String?
    let hospitalAddress: String?
    let treatmentId: String?
    let treatmentName: String?
    let treatmentDescription: String?
    let createdById: String?
    let createdByName: String?
    let reason: String?
    let notes: String?
    let appointmentDate: String?
    let startTime: String?
    let endTime: String?
    let departmentId: String?
    let status: String?
    let createdAt: String?
    let updatedAt: String?
    // Pre-check-in state, set by POST /me/patient/appointments/{id}/pre-checkin.
    let preCheckedIn: Bool?
    let preCheckinTimestamp: String?

    /// Convenience: display name of the doctor/staff
    var doctorName: String? {
        staffName
    }

    /// Formatted time range, e.g. "08:04 – 08:34"
    var timeRange: String? {
        guard let s = startTime, let e = endTime else { return startTime }
        let fmt: (String) -> String = { t in String(t.prefix(5)) } // "08:04:00" → "08:04"
        return "\(fmt(s)) – \(fmt(e))"
    }

    var statusDisplay: String {
        status?.capitalized ?? "Unknown"
    }

    var statusColor: String {
        switch status?.uppercased() {
        case "SCHEDULED", "CONFIRMED": "green"
        case "PENDING": "yellow"
        case "CANCELLED", "NO_SHOW": "red"
        case "COMPLETED": "gray"
        default: "blue"
        }
    }
}

// MARK: - Cancel / Reschedule requests

struct CancelAppointmentRequest: Encodable {
    let appointmentId: String
    let reason: String?
}

struct RescheduleAppointmentRequest: Encodable {
    let appointmentId: String
    let newDate: String
    let newStartTime: String
    let newEndTime: String
    let reason: String?
}

// MARK: - Book Appointment request (POST /me/patient/appointments)

/// PortalBookAppointmentRequestDTO. The backend checks the registration,
/// that the department belongs to the hospital, and assigns the first
/// available provider when staffId is nil; endTime defaults to
/// startTime + 30 min. reason <= 500, notes <= 1000.
struct BookAppointmentRequest: Encodable {
    let hospitalId: String
    let departmentId: String
    let staffId: String?
    let date: String // yyyy-MM-dd
    let startTime: String // HH:mm
    let endTime: String? // HH:mm, nil = server default
    let reason: String?
    let notes: String?
}

// MARK: - Booking wizard lists (GET /me/patient/booking/…)

/// Where the patient holds an active registration.
struct BookingHospitalDTO: Decodable, Identifiable {
    let id: String
    let name: String?
    let address: String?
}

struct BookingDepartmentDTO: Decodable, Identifiable {
    let id: String
    let name: String?
}

struct BookingProviderDTO: Decodable, Identifiable {
    let id: String
    let name: String?
    let fullName: String?
    /// A raw ROLE_* constant with no translation; the sheet shows the name only, as the web does.
    let role: String?

    var displayName: String {
        if let fullName, !fullName.trimmingCharacters(in: .whitespaces).isEmpty { return fullName }
        return name ?? id
    }
}

File: patient-android-app/app/src/main/java/com/bitnesttechs/hms/patient/core/models/AppointmentModels.kt
package com.bitnesttechs.hms.patient.core.models

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass

@JsonClass(generateAdapter = true)
data class AppointmentDto(
    @Json(name = "id") val id: String = "",
    @Json(name = "appointmentDate") val appointmentDate: String = "",
    @Json(name = "startTime") val startTime: String? = null,
    @Json(name = "endTime") val endTime: String? = null,
    @Json(name = "status") val status: String = "",
    @Json(name = "staffId") val staffId: String? = null,
    @Json(name = "staffUserId") val staffUserId: String? = null,
    @Json(name = "staffName") val staffName: String? = null,
    @Json(name = "staffEmail") val staffEmail: String? = null,
    @Json(name = "departmentId") val departmentId: String? = null,
    @Json(name = "departmentName") val departmentName: String? = null,
    @Json(name = "hospitalId") val hospitalId: String? = null,
    @Json(name = "hospitalName") val hospitalName: String? = null,
    @Json(name = "patientId") val patientId: String? = null,
    @Json(name = "notes") val notes: String? = null,
    @Json(name = "reason") val reason: String? = null
) : java.io.Serializable {
    val statusDisplay: String get() = status.replace("_", " ").lowercase()
        .replaceFirstChar { it.uppercase() }
    /** Display time range like "10:00 - 10:30" */
    /** Length of the booked slot; 30 minutes when the record carries no times. */
    val durationMinutes: Long get() {
        val s = startTime?.take(5)?.let { runCatching { java.time.LocalTime.parse(it) }.getOrNull() }
        val e = endTime?.take(5)?.let { runCatching { java.time.LocalTime.parse(it) }.getOrNull() }
        if (s == null || e == null) return 30
        val d = java.time.Duration.between(s, e).toMinutes()
        return if (d in 5..480) d else 30
    }

    val timeDisplay: String? get() {
        val s = startTime?.take(5)
        val e = endTime?.take(5)
        return if (s != null && e != null) "$s - $e" else s
    }
}

/**
 * POST /me/patient/appointments (PortalBookAppointmentRequestDTO). The backend
 * checks the registration, that the department belongs to the hospital, and
 * assigns the first available provider when staffId is null; endTime defaults
 * to startTime + 30 min. reason <= 500, notes <= 1000.
 */
@JsonClass(generateAdapter = true)
data class BookAppointmentRequest(
    @Json(name = "hospitalId") val hospitalId: String,
    @Json(name = "departmentId") val departmentId: String,
    @Json(name = "staffId") val staffId: String? = null,
    @Json(name = "date") val date: String,
    @Json(name = "startTime") val startTime: String,
    @Json(name = "endTime") val endTime: String? = null,
    @Json(name = "reason") val reason: String? = null,
    @Json(name = "notes") val notes: String? = null
)

/** GET /me/patient/booking/hospitals — where the patient holds an active registration. */
@JsonClass(generateAdapter = true)
data class BookingHospitalDto(
    @Json(name = "id") val id: String,
    @Json(name = "name") val name: String? = null,
    @Json(name = "address") val address: String? = null
)

/** GET /me/patient/booking/hospitals/{id}/departments */
@JsonClass(generateAdapter = true)
data class BookingDepartmentDto(
    @Json(name = "id") val id: String,
    @Json(name = "name") val name: String? = null
)

/** GET /me/patient/booking/hospitals/{id}/departments/{id}/providers */
@JsonClass(generateAdapter = true)
data class BookingProviderDto(
    @Json(name = "id") val id: String,
    @Json(name = "name") val name: String? = null,
    @Json(name = "fullName") val fullName: String? = null,
    @Json(name = "role") val role: String? = null
) {
    /** The role is a raw ROLE_* constant with no translation, so the sheet shows the name only, as the web does. */
    val displayName: String get() = fullName?.takeIf { it.isNotBlank() } ?: name.orEmpty()
}

@JsonClass(generateAdapter = true)
data class CancelAppointmentRequest(
    @Json(name = "appointmentId") val appointmentId: String,
    @Json(name = "reason") val reason: String? = null
)

@JsonClass(generateAdapter = true)
data class RescheduleAppointmentRequest(
    @Json(name = "appointmentId") val appointmentId: String,
    @Json(name = "newStartTime") val newStartTime: String? = null,
    @Json(name = "newEndTime") val newEndTime: String? = null,
    @Json(name = "newDate") val newDate: String? = null,
    @Json(name = "reason") val reason: String? = null
)
